src/Nethermind/Nethermind.Core/Threading/ThreadLimiter.cs

```diff
@@ -0,0 +1,46 @@
+// SPDX-FileCopyrightText: 2024 Demerzel Solutions Limited
+// SPDX-License-Identifier: LGPL-3.0-only
+
+using System;
+using System.Threading;
+
+namespace Nethermind.Core.Threading;
+
+/// <summary>
+/// Encapsulate the pattern of checking if new task can be spawned based on a predefined limit.
+/// Used in multithreaded tree visit where we don't know if we can spawn task or not and spawning task itself
+/// is not a cheap operation.
+///
+/// Yes, I don't like the name. Give me a good one.
+/// </summary>
+/// <param name="concurrency">Desired concurrency which include the calling thread. So slot is slot-1.</param>
+public class ThreadLimiter(int concurrency)
+{
+    private int _slots = concurrency - 1;
+
+    public bool TryTakeSlot(out SlotReturner returner)
+    {
+        returner = new SlotReturner(this);
```

Comment:
you always return something that is Disposable, but expect it to be Disposed only if this was `true`, maybe make it nullable and return `null` if this is false?

Reply:
Well, problem is `SlotReturner` is a struct, making it a nullable struct is heavier, and making it reference type is probably even heavier as now its in heap. So it should not make a differece except for having a condition which would probably mess with cpu branch predictor anyway.